parcels/particleset.py

```diff
@@ -89,65 +77,28 @@ def __init__(
         lat=None,
         depth=None,
         time=None,
-        repeatdt=None,
         lonlatdepth_dtype=None,
         pid_orig=None,
-        interaction_distance=None,
-        periodic_domain_zonal=None,
         **kwargs,
     ):
         self.particledata = None
-        self._repeat_starttime = None
-        self._repeatlon = None
-        self._repeatlat = None
-        self._repeatdepth = None
-        self._repeatpclass = None
-        self._repeatkwargs = None
         self._kernel = None
         self._interaction_kernel = None
 
         self.fieldset = fieldset
         self._pclass = pclass
 
-        # ==== first: create a new subclass of the pclass that includes the required variables ==== #
-        # ==== see dynamic-instantiation trick here: https://www.python-course.eu/python3_classes_and_type.php ==== #
-        class_name = pclass.__name__
-        array_class = None
-        if class_name not in dir():
-
-            def ArrayClass_init(self, *args, **kwargs):
-                fieldset = kwargs.get("fieldset", None)
-                ngrids = kwargs.get("ngrids", None)
-                if type(self).ngrids.initial < 0:
-                    numgrids = ngrids
-                    if numgrids is None and fieldset is not None:
-                        numgrids = fieldset.gridset_size
-                    assert numgrids is not None, "Neither fieldsets nor number of grids are specified - exiting."
-                    type(self).ngrids.initial = numgrids
-                self.ngrids = type(self).ngrids.initial
-                if self.ngrids >= 0:
-                    self.ei = np.zeros(self.ngrids, dtype=np.int32)
-                super(type(self), self).__init__(*args, **kwargs)
-
-            array_class_vdict = {
-                "ngrids": Variable("ngrids", dtype=np.int32, to_write=False, initial=-1),
-                "ei": Variable("ei", dtype=np.int32, to_write=False),
-                "__init__": ArrayClass_init,
-            }
-            array_class = type(class_name, (pclass,), array_class_vdict)
-        else:
-            array_class = locals()[class_name]
-        # ==== dynamic re-classing completed ==== #
-        _pclass = array_class
-
         lon = np.empty(shape=0) if lon is None else convert_to_flat_array(lon)
         lat = np.empty(shape=0) if lat is None else convert_to_flat_array(lat)
 
         if isinstance(pid_orig, (type(None), bool)):
             pid_orig = np.arange(lon.size)
 
         if depth is None:
-            mindepth = self.fieldset.dimrange("depth")[0]
+            mindepth = 0
+            for field in self.fieldset.fields.values():
+                if field.grid.depth is not None:
+                    mindepth = min(mindepth, field.grid.depth[0])
```

Comment:
Perhaps not an issue we want to fix now, but for meaningful usecases (“release particles at the surface when no depth is provided”) this assumes that depth is positive downward. What if depth is positive upward? Perhaps make a small TODO note?
```suggestion
                    mindepth = min(mindepth, field.grid.depth[0]). # TODO also support depth-positive-upward grids
```

Reply:
> What if depth is positive upward?

In both instances is the 0 point the ocean surface? In which case, I think it would be beneficial of us to just choose one and assert in the grid initialiser (that way we can make an assumption one way or the other in the rest of the code - and also it's easy to make this change before initialisation using xarray (either manually by users, or automatically by our classmethods that would flow into here).

Reply:
Agree, let's resolve this conversation for now